valuable/src/pointer.rs

```diff
@@ -0,0 +1,196 @@
+//! visit_pointer!(value.x, visitor);   // "a"
+//! visit_pointer!(value.y, visitor);   // Struct2 { field: "b" }
+//! visit_pointer!(value.y.z, visitor); // "b"
+//! ```
+
+use crate::{NamedValues, Slice, Valuable, Value, Visit};
+
+/// A pointer to the value.
+#[derive(Debug, Clone, Copy)]
+#[non_exhaustive]
+pub struct Pointer<'a> {
+    path: &'a [Segment<'a>],
+}
+
+impl<'a> Pointer<'a> {
+    /// Creates a new `Pointer`.
+    pub const fn new(path: &'a [Segment<'a>]) -> Self {
+        Self { path }
+    }
+
+    /// Returns a path pointed by this pointer.
+    ///
+    /// If this pointer points to the current value, this method returns an empty slice.
+    pub fn path(self) -> &'a [Segment<'a>] {
+        self.path
+    }
+
+    #[doc(hidden)]
+    #[must_use]
+    pub fn step(self) -> Self {
+        Self::new(&self.path[1..])
+    }
+}
+
+impl<'a> From<&'a [Segment<'a>]> for Pointer<'a> {
+    fn from(path: &'a [Segment<'a>]) -> Self {
+        Self::new(path)
+    }
+}
+
+/// A segment of a path.
+#[derive(Debug, Clone, Copy)]
+#[non_exhaustive]
+pub enum Segment<'a> {
+    /// Access of a named struct field.
+    Field(&'a str),
+    /// Access of an unnamed struct or a tuple field.
+    TupleIndex(usize),
+    /// Indexing of a list or a map.
+    Index(Value<'a>),
+}
+
+pub(crate) fn visit_pointer<V>(value: &V, pointer: Pointer<'_>, visit: &mut dyn Visit)
+where
+    V: ?Sized + Valuable,
+{
+    let value = value.as_value();
+    if pointer.path.is_empty() {
+        visit.visit_value(value);
+        return;
+    }
+
+    let visitor = &mut Visitor {
+        pointer,
+        visit,
+        index: 0,
+    };
+    match (value, pointer.path[0]) {
+        (Value::Listable(l), Segment::Index(..)) => {
+            l.visit(visitor);
+        }
+        (Value::Mappable(m), Segment::Index(..)) => {
+            m.visit(visitor);
+        }
+        (Value::Tuplable(t), Segment::TupleIndex(..)) => {
+            t.visit(visitor);
+        }
+        (Value::Structable(s), Segment::TupleIndex(..)) if s.definition().fields().is_unnamed() => {
+            s.visit(visitor);
+        }
+        (Value::Structable(s), Segment::Field(..)) if s.definition().fields().is_named() => {
+            s.visit(visitor);
+        }
+        (_, p) => {
+            panic!("invalid pointer: {:?},", p)
+        }
+    }
+}
+
+struct Visitor<'a> {
+    pointer: Pointer<'a>,
+    visit: &'a mut dyn Visit,
+    index: usize,
+}
+
+impl Visit for Visitor<'_> {
+    fn visit_value(&mut self, value: Value<'_>) {
+        if let Segment::Index(index) = self.pointer.path[0] {
+            if index.as_usize() == Some(self.index) {
+                value.visit_pointer(self.pointer.step(), self.visit);
+            }
+        }
+        self.index += 1;
+    }
+
+    fn visit_named_fields(&mut self, named_values: &NamedValues<'_>) {
+        if let Segment::Field(name) = self.pointer.path[0] {
+            if let Some(value) = named_values.get_by_name(name) {
```

Comment:
The challenge will be to avoid using `get_by_name` here in favor of using a field as that would be faster. I'm not sure how we could do it though.

Part of the challenge is that there is no guarantee that the type will be the same each time the pointer is used. It is possible that we could do some sort of caching for faster lookup, but we could punt.

Reply:
I optimized the cases of slice index (https://github.com/tokio-rs/valuable/pull/59/commits/32df15fd06df95c7a2e026ce5ee91f9408f999dd) and structs (https://github.com/tokio-rs/valuable/pull/59/commits/acf64f7bebc9c22028115a2bcaf9c319f37d9bd1).

I don't have any good ideas about optimizing other things.